### ISSUE TITLE
RemoteIp middleware trusted proxies config does not affect Rack::Request::trusted_proxy?

### ISSUE THREAD

**Issue by courtland** [rails/rails#5223]
I have clients that connect to my rails app from a private IP address, where the clients real IP is presented to rails via the X-Forwarded-For header (Thin cluster behind a pound load balancer). By default, rails assumes these private addresses are "trusted proxies", which causes Request#remote_ip to return '127.0.0.1'.

This was somewhat addressed by pull request #2632 by making TRUSTED_PROXIES configurable in the RemoteIp Class.

However, Rails::Rack::Logger methods still call Rack::Request#ip which causes the IP address displayed in my logs to be 127.0.0.1.

Based on this change in rack: https://github.com/rack/rack/pull/192
It seems that ActionDispatch::Request should override Rack::Request#trusted_proxy? with the same "trusted proxies" that are configured for RemoteIp.

At the moment I have the following in an initializer to fix the problem for me, but it is obviously a really bad hack.

```
module Rack
  class Request
    def trusted_proxy?(ip)
      ip =~ /^127\.0\.0\.1$/
    end
  end
end
```

Does anyone have comments or suggestions otherwise? I can attempt a patch if my logic seems sound. Thanks.

**Comment by sockmonk** [comment]
This also seems broken to me on 4.1.6. With default config, it logs all requests (internal and external) as coming from loadbalancers, which have "public" ip addresses. I set config.action_dispatch.ip_spoofing_check = false, and included my internal load balancer ip's in the config.action_dispatch.trusted_proxies option, but it still doesn't solve the problem for me. Just setting the ip_spoofing_check changes nothing. Setting the trusted_proxies setting changes nothing. Doing those things plus monkey-patching ActionDispatch::Request::trusted_proxy? to look at the configured trusted_proxies (as described earlier this thread) causes it to log all requests as 127.0.0.1 instead of the load balancer ips. In my unicorn.log, I can see the actual external ips getting logged in addition to my load balancer ip's.

**Comment by edslocomb** [comment]
For those having problems, which HTTP headers are your trusted proxies / load balancers setting to indicate the upstream client IP?  remote_ip() looks at HTTP_X_FORWARDED_FOR and HTTP_CLIENT_IP (and breaks if both are set, see #15568)

**Comment by leonko** [comment]
Oh, it seems like problem with logger, because I have request.remote_ip == X-Forwarded-For, but logger (    config.log_tags = [:remote_ip] ) anyway print localhost address

**Comment by ppoektos** [comment]
I'm using redmine 3.x + nginx + unicorn. I get 127.0.0.1 in my production log:
```
Started GET "/" for 127.0.0.1 at 2015-08-20 16:33:30 +0200
```
However all headers are correctly sent from nginx to unicorn (192.168.5.1 - it's a client ip):
```
# ps ax | grep 11360
11360 ?        Sl     0:16 unicorn_rails worker[0] -c config/unicorn.rb -E production -D 
# strace -p 11360 -f -e trace=network -s 10000
Process 11360 attached with 2 threads
[pid 11360] accept4(8, 0, NULL, SOCK_CLOEXEC) = 10
[pid 11360] recv(10, "GET / HTTP/1.0\r\nHost: thin\r\nConnection: close\r\nUser-Agent: Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:36.0) Gecko/20100101 Firefox/36\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\r\nAccept-Language: en-us,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nReferer: http://redmine.com/my/page\r\nCookie: _redmine_session=...", 16384, MSG_DONTWAIT) = 1141
[pid 11360] shutdown(10, SHUT_RDWR)     = 0
[pid 11360] accept4(8, 0, NULL, SOCK_CLOEXEC) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11360] accept4(8, 0, NULL, SOCK_CLOEXEC) = 10
[pid 11360] recv(10, "GET /images/news.png HTTP/1.0\r\nX-Forwarded-For: 192.168.5.1\r\nX-Forwarded-Proto: http\r\nHost: redmine.com\r\nX-Real-IP: 192.168.5.1\r\nCLIENT_IP: 192.168.5.1\r\nHTTP_CLIENT_IP: 192.168.5.1\r\nConnection: close\r\nUser-Agent: Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:36.0) Gecko/20100101 Firefox/36\r\nAccept: image/png,image/*;q=0.8,*/*;q=0.5\r\nAccept-Language: en-us,en;q=0.5\r\nAccept-Encoding: gzip, deflate\r\nReferer: http://redmine.com/stylesheets/application.css\r\nCookie: _redmine_session=...

```

**Comment by courtland** [comment]
Nearly 4 years later I am finally getting around to upgrading from rails 3.2.x to 4.2.x and I can also confirm that this issue still exists to some extent.

As in my original post, am still using a pound load balancer in front of thin web servers. So now with rails 4.2.5, remote_ip in the log is incorrect (shows 127.0.0.1 instead of e.g., 192.168.5.6) as is Request#remote_ip.

The problem with Request#remote_ip being wrong was caused by my trusted_proxies configuration. I had configured in application.rb:

```ruby
config.action_dispatch.trusted_proxies = /\A127\.0\.0\.1\z/
```
Looking at the code and documentation for the RemoteIP::IpSpoofAttackError class, I noticed that the behavior now changes when the trusted_proxies is an Array vs. a single object (String or Regex). If it is a single object, it is appended to the default TRUSTED_PROXIES array. In my case this caused "192.168.0.0/16" and the other private networks to be detected as proxies, which is not what I want.
Changing my application.rb to specify an Array for trusted_proxies fixed the Request#remote_ip problem:

```ruby
config.action_dispatch.trusted_proxies = %w( 127.0.0.1 )
```

Including remote_ip in config.log_tags was still problematic (logs showed 127.0.0.1 instead of 192.168.5.x). I was able to fix this by including the following in application.rb, after my trusted_proxies configuration line.
```
    config.action_dispatch.trusted_proxies = %w( 127.0.0.1 )
    config.middleware.delete 'ActionDispatch::RemoteIp'
    config.middleware.insert_before Rails::Rack::Logger, ActionDispatch::RemoteIp, config.action_dispatch.ip_spoofing_check, config.action_dispatch.trusted_proxies
```
I learned about the fix from pull request #21798, which seems to also be resolved by pull request #19710. I believe one of these needs to be merged for this problem to be fully resolved.

**Comment by ppoektos** [comment]
Unfortunately this trick doesn't work for me (redmine 3.2, rails 4.2.5, nginx + unicorn) and I'm still getting
```
Started GET "/issues/12345/time_entries?page=1" for 127.0.0.1 at 2016-01-25 14:48:08 +0100

```

**Comment by antaflos** [comment]
As of Rails 4.2.0 trusted proxies are defined as an array of `IPAddr` objects, e.g.:

```ruby
config.action_dispatch.trusted_proxies = %w(127.0.0.1 10.1.2.0/16).map { |proxy| IPAddr.new(proxy) }
```

**Comment by JasonBarnabe** [comment]
I've gone through a few hours of banging my head against this in a Rails 4.0 app. I thought it might be useful to summarize my findings.

Rails considers anything from 172.16.0.0 to 172.31.255.255 (among other things) to be a trusted proxy. This covered all users of my (internal) app. So I had to change the `trusted_proxies`setting. 

As @courtland mentions, the type of object you use for `trusted_proxies` matters. In Rails [4.0](https://github.com/rails/rails/blob/4-0-stable/actionpack/lib/action_dispatch/middleware/remote_ip.rb#L60-L67) and [4.1](https://github.com/rails/rails/blob/4-1-stable/actionpack/lib/action_dispatch/middleware/remote_ip.rb#L60-L67), if you pass in a `RegExp`, this will represent the **full list** of trusted proxies. If you pass any other non-`nil` value, this will represent **additions** to the trusted proxy list. In Rails [4.2](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/middleware/remote_ip.rb#L63-L69), it is an `Array` that represents a full list of trusted proxies, and every other type is an addition.

As @batizhevsky is attempting to fix in [his PR](https://github.com/rails/rails/pull/19710), you must also ensure that the `ActionDispatch::RemoteIp` middleware runs before anything attempts to read `remote_ip` from the request because [`remote_ip` is cached](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/http/request.rb#L231). In my case, I had `log_tags` configured to use `remote_ip`, so I had to switch the order of the middleware around:

```ruby
  config.middleware.delete(ActionDispatch::RemoteIp)
  config.middleware.insert_before(Rails::Rack::Logger, ActionDispatch::RemoteIp, true, trusted_proxies)
```

**Comment by kvokka** [comment]
This bug is still present in Rails 5.0.0.1. Syntax changes a little, so, I had to add in 
    # config/application.rb
    config.action_dispatch.trusted_proxies = %w(127.0.0.1 ::1).map { |proxy| IPAddr.new(proxy) }
    config.middleware.insert_before(Rails::Rack::Logger,
                                    ActionDispatch::RemoteIp,
                                    true,
                                    config.action_dispatch.trusted_proxies)

   # config/initializers/action_dispatch.rb
    module TrustedProxyMonkeyPatch
      def ip
        @ip ||= (get_header("action_dispatch.remote_ip") || super).to_s
      end
    end
    ActionDispatch::Request.send :include, TrustedProxyMonkeyPatch

**Comment by maclover7** [comment]
Looks like there might be a PR (#14600) in progress to fix this cc @matthewd

**Comment by spurnaye** [comment]
Same issues here. 
I am using nginx + unicorn + rails 5.0.1
nginx configuration is same as https://www.digitalocean.com/community/tutorials/how-to-deploy-rails-apps-using-unicorn-and-nginx-on-centos-6-5

Atleast the request.remote_ip issue is resolved by following but logs are still incorrect
in config/application.rb
```
# hopefully, this should not be breaking change if rails version changes!
config.action_dispatch.trusted_proxies = %w(127.0.0.1 ::1).map {
  |proxy| IPAddr.new(proxy) }
```
### Steps to reproduce

(Guidelines for creating a bug report are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

### Expected behavior
The logs should print correct remote ip

### Actual behavior
started GET "/log_in" for 127.0.0.1 at <TIME>

### System configuration
**Rails version**:
5.0.1
**Ruby version**:
ruby 2.3.1p112

**Comment by jonathanhefner** [comment]
> However, Rails::Rack::Logger methods still call Rack::Request#ip which causes the IP address displayed in my logs to be 127.0.0.1.

Fixed by #33489.

> Including remote_ip in config.log_tags was still problematic

Fixed by #27515.

If I missed any problems in this thread, or if I am mistaken about the above problems being fixed, please open a new issue and please include a minimal reproduction script.

**Comment by vanboom** [comment]
Confirming this for Rails 5.2.4.5 

```
    config.action_dispatch.trusted_proxies = %w(127.0.0.1 11.0.0.1).map { |proxy| IPAddr.new(proxy) }
```

Has no effect on the RemoteIp behavior.   11.0.0.1 is treated as a client IP and is not being discarded as a trusted proxy.
